specs/DataStructures/Trie.md

```diff
@@ -0,0 +1,77 @@
+# Trie
+
+We use Merkle-Patricia Trie to store blockchain state.
+Trie is persistent, which means that insertion of new node actually leads to creation of new path to this node, and thus root of Trie after insertion will also be presented by new object.
+
+Here we describe structures closely related to Runtime.
+
+## Main structures
+
+### Trie
+
+Base structure. Almost not called directly from `Runtime`, modifications are made using `TrieUpdate`.
+
+### TrieUpdate
+
+Provides a way to access Storage and record changes to commit in the future. Update is prepared as follows:
+
+- changes are made using `set` and `remove` methods, which are added to `prospective` field,
+- call `commit` method which moves `prospective` changes to `committed`,
+- call `finalize` method which prepares `TrieChanges` and state changes based on `committed` field.
+
+Note that `finalize` does not modify `Trie` itself, it only prepares changes to be passed to `ChainStore`.
```

Comment:
I think what you meant here is `finalize` and `Trie::update` does not update the database storage. It only modifies trie nodes in memory. These changes are written to the database when `ShardTries::apply_insertions` is called